File: examples/typegraphs/http-runtime.py
# Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
# SPDX-License-Identifier: MPL-2.0

# skip:start
from typegraph import Graph, Policy, t, typegraph
from typegraph.graph.params import Cors

# skip:end
# highlight-next-line
from typegraph.runtimes import HttpRuntime


@typegraph(
    # skip:start
    cors=Cors(allow_origin=["https://metatype.dev", "http://localhost:3000"]),
    # skip:end
)
def http_runtime(g: Graph):
    pub = Policy.public()

    # highlight-next-line
    facts = HttpRuntime("https://uselessfacts.jsph.pl/api/v2/facts")

    g.expose(
        pub,
        facts=facts.get(
            "/random",
            t.struct({"language": t.enum(["en", "de"])}),
            t.struct(
                {
                    "id": t.string(),
                    "text": t.string(),
                    "source": t.string(),
                    "source_url": t.string(),
                    "language": t.string(),
                    "permalink": t.string(),
                },
            ),
        ),
        facts_as_text=facts.get(
            "/random",
            t.struct(
                {
                    "header_accept": t.string().set("text/plain"),
                    "language": t.enum(["en", "de"]),
                },
            ),
            t.string(),
            header_prefix="header_",
        ),
    )

File: examples/typegraphs/http-runtime.ts
// skip:start
import { Policy, t, typegraph } from "@typegraph/sdk/index.ts";

// isort: off
// skip:end
// highlight-next-line
import { HttpRuntime } from "@typegraph/sdk/runtimes/http.ts";

await typegraph(
  {
    name: "http-runtime",
    // skip:next-line
    cors: { allowOrigin: ["https://metatype.dev", "http://localhost:3000"] },
  },
  (g) => {
    // highlight-next-line
    const facts = new HttpRuntime("https://uselessfacts.jsph.pl/api/v2/facts");
    const pub = Policy.public();

    g.expose(
      {
        facts: facts.get(
          t.struct({
            language: t.enum_(["en", "de"]),
          }),
          t.struct({
            id: t.string(),
            text: t.string(),
            source: t.string(),
            source_url: t.string(),
            language: t.string(),
            permalink: t.string(),
          }),
          {
            path: "/random",
          }
        ),
        facts_as_text: facts.get(
          t.struct({
            header_accept: t.string().set("text/plain"),
            language: t.enum_(["en", "de"]),
          }),
          t.string(),
          { path: "/random", headerPrefix: "header_" }
        ),
      },
      pub
    );
  }
);
